kadasrouting/core/memorylayersaver.py

```diff
@@ -0,0 +1,365 @@
+            layer.committedAttributesDeleted.disconnect(self.setProjectDirty2)
+            layer.committedAttributesAdded.disconnect(self.setProjectDirty2)
+            if self._deleteSignalOk:
+                layer.committedFeaturesRemoved.disconnect(self.setProjectDirty2)
+            layer.committedFeaturesAdded.disconnect(self.setProjectDirty2)
+            layer.committedAttributeValuesChanges.disconnect(self.setProjectDirty2)
+            layer.committedGeometriesChanges.disconnect(self.setProjectDirty2)
+
+    def connectMemoryLayers( self ):
+        for layer in self.memoryLayers():
+            self.connectProvider( layer )
+
+    def disconnectMemoryLayers( self ):
+        for layer in self.memoryLayers():
+            self.disconnectProvider( layer )
+
+
+    def unload(self):      
+        # self._iface.removePluginMenu("&Test tools",self._loadadjaction)
+        pass
+
+    def loadData(self):
+        filename = self.memoryLayerFile()
+        file = QFile(filename)
+        if file.exists():
+            layers = list(self.memoryLayers())
+            if layers:
+                try:
+                    with Reader(filename) as reader:
+                        reader.readLayers(layers)
+                except:
+                    QMessageBox.information(
+                        self._iface.mainWindow(),"Error reloading memory layers",
+                        str(sys.exc_info()[1]) ) 
+
+    def saveData(self):
+        try:
+            filename = self.memoryLayerFile()
+            try:
+                file=QFile(finfo.filePath())
+                file.remove()
+            except:
+                pass
+            layers = list(self.memoryLayers())
+            if layers:
+                with Writer(filename) as writer:
+                    writer.writeLayers( layers )
+        except:
+            raise
+            QMessageBox.information(self._iface.mainWindow(),"Error saving memory layers",
+                                    str(sys.exc_info()[1]) )
+
+    def memoryLayers(self):
+        for l in list(QgsProject.instance().mapLayers().values()):
+            if self.isSavedLayer(l):
+                yield l
+
+    def isSavedLayer( self, l ):
+        if l.type() != QgsMapLayer.VectorLayer:
+            return
+        pr = l.dataProvider()
+        if (pr is None) or (pr.name() != 'memory'):
+            return False
+        use = l.customProperty("SaveMemoryProvider")
+        return not (use == False)
+
+    def memoryLayerFile( self ):
+        name = QgsProject.instance().fileName()
+        if not name:
+            return ''
+        lname = name+".mldata"
+        return lname
+
+    def clearMemoryProvider(self, lyr):
+        pl = lyr.dataProvider()
+        pl.select()
+        f = QgsFeature()
+        while pl.nextFeature(f):
+            pl.deleteFeatures(f.id())
+        pl.deleteAttributes(pl.attributeIndexes())
+
+    def setProjectDirty2(self,value1,value2):
```

Comment:
These can be probably be put in a single method

Reply:
won't fix